### ISSUE TITLE
Storm Drain Add Inlet Type 4 data

### ISSUE THREAD

**Issue by FLO-2DJJ** [FLO-2DSoftware/qgis-flo-2d-plugin#1088]
In the SD Widget Editor the option to add inlet type 4 should consider 2 options: rating table (currently the default) but also a Culvert Equation.

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/c305d323-f02b-4045-9e4d-a5f7c05fda7a)

**Comment by FLO-2DNoemi** [comment]
JJ,

It is not working for me  but Karen was testing this issue. She may have
more information on this issue.


Noemi

**Comment by FLO-2DKaren** [comment]
@FLO-2DNoemi 
Noemi, 
You couldn't import the culvert for a type 4?  It has to have a file with the data just like a single file for a rating table.  The name of the file should match the inlet.

**Comment by FLO-2DKaren** [comment]
OK let's close 1088 and do the work on #1052 since it has the details and this one is just confusing.

**Comment by FLO-2DKaren** [comment]
@FLO-2DJJ 

JJ, Can you look at my video and see what I'm doing wrong?

https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424460/6158e7c4-881d-4b9c-8c27-d536b203efb3

**Comment by FLO-2DKaren** [comment]
Here's the project I'm using.
https://flo-2d.sharefile.com/d-s6bc5a367272b4f308680c1cda40df030

**Comment by FLO-2DJJ** [comment]
OK Karen, I'll look at it.

Juan Jose Rodriguez
FLO-2D Software
www.flo-2d.com
***@***.***

**Comment by FLO-2DJJ** [comment]
Hi @FLO-2DKaren,
The way to associate Culvert equations to inlets is thru a file named "Type4Culvert.txt" with a line for each inlet with Culvert eq.:

cell  inlet_name  CDDIAMETR TYPEC TYPEEN CUBASE MULTBARRELS
(A cell number should be there but is not used).

In your example with only one inlet, Type4Culvert.txt could contain:
999 I4-37-32-26-1   4  1  1  4  2

The file is read using the read tool:
![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/0157eff6-8469-41b2-a5cf-6c9edea2038f)

@FLO-2DNoemi and I reached to this mechanism after considering several options. We where not completely pleased with this one but the other alternatives were even more convoluted.

Similarly, to associate a rating table with an inlet a file with the same name as the inlet is used.

Hope this helps you.

**Comment by FLO-2DNoemi** [comment]
I tested it and this is not working properly. Even the import file option
seems to be broken for type 4 generalized culvert equation.

**Comment by FLO-2DJJ** [comment]
I tested it again and again, and works. It works importing Rating Tables and/or Culvert eq., or a combination of both.

@FLO-2DNoemi are you using the latest master from GitHub? If so, what steps give you the wrong results?

**Comment by FLO-2DKaren** [comment]
JJ importing isn’t the problem. It’s building the data with the new
editors.

Karen O'Brien
Technical Support

FLO-2D Software | Riada Engineering
P.O. Box 66 | Nutrioso, AZ | 85932
T. 928.339.1935
www.flo-2d.com | ***@***.***

**Comment by FLO-2DJJ** [comment]
Let's clarify this.
The new options 'Add Rating Table' and 'Add Culvert Equation' allow to add and edit Type 4 conditions:
![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/09a50ab3-00f2-4d37-890a-05b4ce9aee5d)

BUT (a big 'BUT'), for the time being, they can not be assigned to any inlet BECAUSE the mechanism to do the assigning is by importing a file Type4Culvert.txt (for Culvert eqs.) and/or individual files with the same name of the inlets for rating tables.

Years ago the user could select a type 4 data (RT or Culvert) in the dropdown list of the inlets dialog. But it is disabled now due to the assignment mechanism described above. 

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/e8c50b9b-3daf-4a27-9d72-0e44cf938fb6)

The dropdown list selection created lots of problems in the past due to the idea of using prefixes for inlet names and RT and Culvert eq, and the type of the inlet. There were also problems with the selection of the same RT or Culvert eq to more than one inlet. An other issues that I can not recall now. 

@FLO-2DNoemi, this is something we discussed a lot and concluded to do it this way. I as said before, we were not completely pleased with this solution.

**Comment by FLO-2DJJ** [comment]
@FLO-2DKaren, @FLO-2DNoemi 
A way around this is to extend the functionality of 'Add Rating Table' and 'Add Culvert Equation' by allowing the user to change the name 'RatingTableN' and 'CulvertEqN' (the default names) to an existing inlet name of type 4.
The above could be done either when the the table is created 

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/7cf02e63-3626-4dda-9797-25305e7ab1c3)

or when using the 'Rename inlet Type 4 condition' tool:

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/fca7385e-17bf-4235-8424-49cda59ab07a)

Either way we should include the same checks and confirmations used when importing Type4Culvert.txt and individual files with the same name as inlets.

Any other approach or ideas @FLO-2DNoemi, @FLO-2DKaren?

**Comment by FLO-2DKaren** [comment]
Why don't you review the Structure Editor since it already works the way we want this one to.

**Comment by FLO-2DJJ** [comment]
The hydraulic structures editor has different functionality. Storm drain
has its own logic and complexities. It can not be modeled using the
hydraulic structures.

**Comment by FLO-2DJJ** [comment]
@FLO-2DKaren, @FLO-2DNoemi. 
Let me rephrase what I said before. Yes Karen, you are right, the hydraulic structures editor is a good guide to do the rating tables/culvert equations assignments for storm drain. In fact, it already does it when importing the data. We only need to allow the name assignment when new tables/equations are added with the “Add…” tools (that the hydraulic structures editor doesn’t have):

![image](https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/assets/20424575/f149a09b-a8dc-4af6-a49c-5fedcf402863)

**Comment by FLO-2DJJ** [comment]
@FLO-2DKaren, @FLO-2DNoemi 
The master repo in GitHub has functionality to add RT/Culvert eq. and rename them with the same name of existing inlets. 
Several checks are done before performing the rename and assignment to maintain consistency with the logic of the inlet types.
